balancer/balancer.go

```diff
@@ -371,3 +376,21 @@ type ClientConnState struct {
 // ErrBadResolverState may be returned by UpdateClientConnState to indicate a
 // problem with the provided name resolver data.
 var ErrBadResolverState = errors.New("bad resolver state")
+
+// A ProducerBuilder is a simple constructor for a Producer.  It is used by the
+// SubConn to create producers when needed.
+type ProducerBuilder interface {
+	// Build creates a Producer.  The first parameter is always a
+	// grpc.InvokerStreamer (a type to allow creating RPCs/streams on the
+	// associated SubConn), but is declared as interface{} to avoid a
+	// dependency cycle.  Should also return a close function that will be
+	// called when all references to the Producer have been given up.
+	Build(grpcInvokerStreamer interface{}) (p Producer, close func())
+}
+
+// A Producer is a type shared among potentially many consumers.  It is
+// associated with a SubConn, and an implementation will typically contain
+// other methods to provide additional functionality, e.g. configuration or
+// subscription registration.
```

Comment:
This description feels very vague to me, in the sense that I can't get a good idea of when to use a `Producer`, how to use one etc.

Reply:
> Additional functionality with respect to what? There's no functionality defined before this.

In addition to the builder that created it, I suppose, and the `close` function the builder returns.

> Do we expect this to ever be implemented by another concrete type?

Yes, definitely.  This is intended to be generic and able to be implemented by anything.  The other example we already have is the health checks, which are implemented in the other languages as a generic wrapper, but hard-coded into gRPC in Go.  I'd like to move to a similar model there by using this mechanism.

> This comment is very specific to type producer in orca/producer.

Not at all.  If it was, it would talk about ORCA configuration.  ORCA should not be part of grpc-go directly, but instead should only be a plugin which uses this.